miden/tests/integration/flow_control/mod.rs

```diff
@@ -352,3 +355,64 @@ fn simple_dyncall() {
         false,
     );
 }
+
+// PROCREF INSTRUCTION
+// ================================================================================================
+
+#[test]
+fn procref() {
+    let source = "
+    use.std::math::u64
+
+    proc.foo.4
+        push.3.4
+    end
+
+    begin
+        procref.u64::overflowing_add
+        push.0
+        procref.foo
+    end";
+
+    let assembler = Assembler::default();
+    let program_ast = ProgramAst::parse(source).unwrap();
+
+    // compile program to get mast roots from the CodeBlock
+    let compiled_program = assembler
+        .with_library(&StdLibrary::default())
+        .unwrap()
+        .compile_in_context(&program_ast, &mut AssemblyContext::for_program(Some(&program_ast)))
+        .unwrap();
+
+    let mut mast_roots = Vec::new();
+    let span = match compiled_program {
+        CodeBlock::Span(span) => span.clone(),
+        _ => unreachable!(),
+    };
+    for batch in span.op_batches() {
+        for op in batch.ops() {
+            match op {
+                // procref uses only `Push` operations
+                vm_core::Operation::Push(v) => mast_roots.push(v.as_int()),
+                _ => {}
+            }
+        }
+    }
```

Comment:
I'm not sure this is testing the right thing. Basically, you are getting MAST roots from the `procref` instructions and then checking against the same instructions below - so, they will always match, even if the roots are not correct.

I'd try to use `compile_module()` function to get `ModuleAst` and then get procedure roots from there.

Reply:
Hmm, indeed, sorry. Although I think it will be easier not to try to generate procedure roots from `ModuleAst` (and it could be tricky in case of local procs since we can't compile program as module), but to get it from assembler and assembly context, where roots will be already generated.

Reply:
I would prefer not to expose `proc_cache` publicly. It shouldn't be difficult to get procedure roots by using `compile_module()` to do this you can:

1. Crete a module with the procedures you are interested in (all exported).
2. Compile it using `compile_module()`.
3. Associated the list of returned digests with the procedures - i.e., the first returned digests is for the first procedure etc.

Reply:
Also, you don't need to compile program module for this - just duplicate the procedure in the library module. Procedure with the same body should have the same MAST roots.

Reply:
Got it. I'll update it right away